airflow/utils/helpers.py

```diff
@@ -244,3 +244,18 @@ def build_airflow_url_with_query(query: Dict[str, Any]) -> str:
     view = conf.get('webserver', 'dag_default_view').lower()
     url = url_for(f"Airflow.{view}")
     return f"{url}?{parse.urlencode(query)}"
+
+
+def exactly_one(*args):
+    """
+    Returns True if exactly one of *args is "truthy", and False otherwise.
+    :param args:
+    :return:
+    """
+    has_one = False
+    for arg in args:
+        if arg and has_one:
+            return False
+        elif arg:
+            has_one = True
+    return has_one
```

Comment:
This can be problematic if the argument is of a type that can accept a falsy value, or if the default is not falsy (e.g. `NOTSET`). This should Can accept an optional `check_default=` keyword argument and check for default-ness with that instead (the default would be e.g. `functools.partial(operator.is_not, None)`).

Reply:
My intention is that the user must provide args that are truthy/falsy  

So in that case you would do 

```python
exactly_one(a != NOTSET, b != NOTSET)

```

etc.... what do you think?

Reply:
so the meaning is, "exactly one is truthy", not "exactly one is set"

Reply:
exploring that thought...

```python
def exactly_one(*args, func=bool):
    """
    Returns True if exactly one of *args is "truthy", and False otherwise.
    :param args:
    :return:
    """
    return sum(map(func, args)) == 1

exactly_one(a, b, c, func=lambda x: x != NOTSET)
```

Given that you could achieve the same result, with the simpler function in this way:

```python
exactly_one(*map(lambda x: x !=NOTSET, vals))
```

I think maybe keeping the function simpler is better?

Reply:
That would work. In that case though, I’d make some of the existing changes more explicit e.g. `exactly_one(execution_date is not None, run_id is not None)` (that’s a bit verbose to me though, so parametrising `exactly_one` should still be considered IMO), and document the function in detail so future contributors don’t fall in this pitfall. On second thought, this falsy-none problem may be an unnecessary pitfall in the first place, indicating we should build safeguards in `exactly_one` in the first place.

Reply:
>  That would work. In that case though, I’d make some of the existing changes more explicit e.g. exactly_one(execution_date is not None, run_id is not None)

I went with `exactly_one(execution_date, run_id)` because AFAIK those values obey truthy evaluation.  Are you suggesting that we do `is not None` because we should never do truthy evaluation?  Or because in this particular case there is a reason not to?  When it was XOR there was a reason (namely because XOR doesn't handle truthy), but seems here it should be OK,  and as you pointed out it's less verbose -- wdyt?

> that’s a bit verbose to me though, so parametrising exactly_one should still be considered IMO

I did explore the idea above, and arrived at the feeling that it wasn't worth it.  But if you feel strongly we should parameterize please say so and perhaps propose a manner.  It doesn't really cost too much to parameterize just makes it a little more complicated to understand the behavior.

> On second thought, this falsy-none problem may be an unnecessary pitfall in the first place, indicating we should build safeguards in exactly_one in the first place.

can you clarify?

Reply:
> I went with `exactly_one(execution_date, run_id)` because AFAIK those values obey truthy evaluation. Are you suggesting that we do `is not None` because we should never do truthy evaluation? Or because in this particular case there is a reason not to?

Because `run_id` can be an empty string, and we want to show a different error for this situation.

> can you clarify?

Basically the above. The main purpose of this function (well, the xor checks it is replacing) is to check whether arguments are passed, not whether those arguments are falsy. And if you don’t either explicitly use `is not None`, or include stricter evaluation logic in `exactly_one`, this creates a subtle semantic difference between the caller site and the body of `exactly_one` that can potentially trip people up and cause subtle, unintended behavioural change. And this situation is complete avoidable by changing the evaluation logic to not rely on falsiness.

Reply:
> Because run_id can be an empty string, and we want to show a different error for this situation.

OK in that case I updated XCom to use `is not None` to ensure no change in behavior.  And out of an abundance of caution, same for emr add steps.  The others did not need it.

> subtle, unintended behavioural change

Yeah I was mindful of the potential that it would be a meaningful change, and did consider leaving the xcom calls with `is not None` to begin with.  But I wanted to put forward the less verbose approach with the expectation that we'd confirm through the PR process whether it's  kosher (as we've now done :) ).

So, there should now be no breaking changes in this PR, and what remains is to decide on whether we should parameterize or not and if so how.  It's also something we could add later.